platform/xen/rumpxendev/evtdev.c

```diff
@@ -0,0 +1,502 @@
+
+	/* give us a rump kernel context */
+	rumpuser__hyp.hyp_schedule();
+	rumpuser__hyp.hyp_lwproc_newlwp(0);
+	rumpuser__hyp.hyp_unschedule();
+
+	for (;;) {
+		minios_wait_event(minios_events_waitq, xenevt_ring_prod != prod);
+		prod = xenevt_ring_prod;
+		cons = xenevt_ring_cons;
+
+		membar_sync();
+
+		while (cons != prod) {
+			u_int port = xenevt_ring[cons];
+			struct xenevt_dev_data *d;
+
+			KASSERT(port < NR_EVENT_CHANNELS);
+
+			mutex_enter(&devevent_lock);
+
+			d = devevents[port];
+
+			KASSERT(d);
+
+			mutex_enter(&d->lock);
+
+			queue(d, port);
+
+			mutex_exit(&d->lock);
+			mutex_exit(&devevent_lock);
+
+			cons++;
+		}
+
+		membar_sync();
+
+		xenevt_ring_cons = cons;
+	}
+}
+
+/*----- request handling (writes to the device) -----*/
+static int
+xenevt_dev_write(struct file *fp, off_t *offset, struct uio *uio,
+		 kauth_cred_t cred, int flags)
+{
+	struct xenevt_dev_data *d = fp->f_data;
+	uint16_t *chans = NULL;
+	int i, nentries, err;
+
+	DPRINTF(("/dev/xenevt: write...\n"));
+
+	if (uio->uio_resid == 0) {
+		err = 0;
+		goto out;
+	}
+
+	nentries = uio->uio_resid / sizeof(uint16_t);
+	if (nentries > NR_EVENT_CHANNELS) {
+		err = EMSGSIZE;
+		goto out;
+	}
+
+	chans = bmk_memcalloc(nentries, sizeof(uint16_t));
+	if (chans == NULL) {
+		err = ENOMEM;
+		goto out;
+	}
+
+	err = uiomove(chans, uio->uio_resid, uio);
+	if (err) goto out;
+
+	mutex_enter(&devevent_lock);
+	for (i = 0; i < nentries; i++) {
+		if (chans[i] < NR_EVENT_CHANNELS &&
+		    devevents[chans[i]] == d)
+			minios_unmask_evtchn(chans[i]);
+	}
+	mutex_exit(&devevent_lock);
+
+	err = 0;
+out:
+	DPRINTF(("/dev/xenevt: write done, err=%d\n", err));
+	bmk_memfree(chans);
+	return err;
+}
+
+static int
+xenevt_dev_read(struct file *fp, off_t *offset, struct uio *uio,
+		kauth_cred_t cred, int read_flags)
+{
+	struct xenevt_dev_data *d = fp->f_data;
+	u_int cons, prod, len, uio_len;
+	int err;
+
+	DPRINTF(("/dev/xenevt: read...\n"));
+
+	mutex_enter(&d->lock);
+
+	err = 0;
+	while (err == 0) {
+		cons = d->ring_cons;
+		prod = d->ring_prod;
+
+		if (cons != prod) break; /* data available */
+
+		if (d->flags & EVTDEV_F_OVERFLOW) break;
+
+		/* nothing to read */
+		if ((fp->f_flag & FNONBLOCK) == 0)
+			err = cv_wait_sig(&d->cv, &d->lock);
+		else
+			err = EAGAIN;
+	}
+
+	if (err == 0 && (d->flags & EVTDEV_F_OVERFLOW))
+		err = EFBIG;
+
+	mutex_exit(&d->lock);
+
+	if (err) goto out;
+
+	uio_len = uio->uio_resid / BYTES_PER_PORT;
+	if (cons <= prod)
+		len = prod - cons;
+	else
+		len = EVTDEV_RING_SIZE - cons;
+	if (len > uio_len)
+		len = uio_len;
+	err = uiomove(&d->ring[cons], len * BYTES_PER_PORT, uio);
+	if (err) goto out;
+
+	cons = (cons + len) & EVTDEV_RING_MASK;
+	uio_len = uio->uio_resid / BYTES_PER_PORT;
+	if (uio_len == 0) goto done;
+
+	/* ring wrapped */
+	len = prod - cons;
+	if (len > uio_len)
+		len = uio_len;
+	err = uiomove(&d->ring[cons], len * BYTES_PER_PORT, uio);
+	if (err) goto out;
+	cons = (cons + len) & EVTDEV_RING_MASK;
+
```

Comment:
This bit looks suspicious.  What's preventing two callers from racing into the read function?

Granted, the rules are probably a bit different since this is a unikernel, and that does not constitute a security problem, but I'd still like to see all code kept "kernel quality".